source/slang/slang-emit-c-like.cpp

```diff
@@ -2019,18 +2057,35 @@ void CLikeSourceEmitter::defaultEmitInstExpr(IRInst* inst, const EmitOpInfo& inO
 
         IRFieldAddress* ii = (IRFieldAddress*) inst;
 
-        auto prec = getInfo(EmitOp::Postfix);
-        needClose = maybeEmitParens(outerPrec, prec);
-
-        auto base = ii->getBase();
-        emitOperand(base, leftSide(outerPrec, prec));
-        m_writer->emit(".");
-        if(getSourceLanguage() == SourceLanguage::GLSL
-            && as<IRUniformParameterGroupType>(base->getDataType()))
+        if (supportPtrTypes())
         {
-            m_writer->emit("_data.");
+            auto prec = getInfo(EmitOp::Prefix);
+            needClose = maybeEmitParens(outerPrec, prec);
+            m_writer->emit("&");
+            outerPrec = rightSide(outerPrec, prec);
+            auto innerPrec = getInfo(EmitOp::Postfix);
+            bool innerNeedClose = maybeEmitParens(outerPrec, innerPrec);
+            auto base = ii->getBase();
+            emitOperand(base, leftSide(outerPrec, innerPrec));
+            m_writer->emit("->");
+            m_writer->emit(getName(ii->getField()));
+            maybeCloseParens(innerNeedClose);
+        }
+        else
+        {
+            auto prec = getInfo(EmitOp::Postfix);
+            needClose = maybeEmitParens(outerPrec, prec);
+
+            auto base = ii->getBase();
+            emitOperand(base, leftSide(outerPrec, prec));
+            m_writer->emit(".");
+            if(getSourceLanguage() == SourceLanguage::GLSL
+                && as<IRUniformParameterGroupType>(base->getDataType()))
+            {
+                m_writer->emit("_data.");
+            }
+            m_writer->emit(getName(ii->getField()));
```

Comment:
Some of this could conceivably be simplified by treating this as emitting `&((*base).field)` and then using subroutines like `emitDereferenceOperand` to emit both the `*` and `&`. For a pointer-based target both of those would emit, and for a non-pointer target neither would.

Of course that approach wouldn't get you the nice `->` instead of `(*base).field`, but we need to strike a balance here between the clarity of the generated code and the clarity of the emit logic.

Reply:
Done.